devices/pine64-pinetab/u-boot/default.nix

```diff
@@ -0,0 +1,110 @@
+{ lib
+, writeText
+, callPackage
+, buildUBoot
+, armTrustedFirmwareAllwinner
+, crustFirmware
+, fetchpatch
+, fetchurl
+, fetchFromGitHub
+, buildPackages
+}:
+
+let
+  pw = id: sha256: fetchpatch {
+    inherit sha256;
+    name = "${id}.patch";
+    url = "https://patchwork.ozlabs.org/patch/${id}/raw/";
+  };
+
+  pine64Patch = rev: sha256: fetchpatch {
+    inherit sha256;
+    name = "pine64-${rev}.patch";
+    url = "https://gitlab.com/pine64-org/u-boot/-/commit/${rev}.patch";
+  };
+
+  # use a version of ATF that has all the Crust goodies in it
+  crustATF = armTrustedFirmwareAllwinner.overrideAttrs(old: rec {
+    name = "arm-trusted-firmware-crust-${version}";
+    version = "2.4";
+    src = fetchFromGitHub {
+      owner = "crust-firmware";
+      repo = "arm-trusted-firmware";
+      rev = "42b9ab0cbe6c1d687fe331c547d28489e12260c3";
+      sha256 = "13q0946qk2brda1ci3bsri359ly8zhz76f2d1svnlh45rrrfn984";
+    };
+  });
+in
+(buildUBoot {
+  defconfig = "pinephone_defconfig";
+  extraMeta.platforms = ["aarch64-linux"];
+  BL31 = "${crustATF}/bl31.bin";
+  SCP = "${crustFirmware}/scp.bin";
+
+  extraPatches = [
+    # usb: Make USB_MUSB_PIO_ONLY selected by USB_MUSB_SUNXI
+    # https://patchwork.ozlabs.org/patch/1202024
+    (pw "1202024" "0c196zk1s3pq3wdv909sxmjgqpll2hwb817bpbghkfkyyknl96vg")
+
+    # sunxi: support asymmetric dual rank DRAM on A64/R40
+    # https://patchwork.ozlabs.org/patch/1312857
+    (pw "1312857" "0w8caf936y7f1r2lvdqar9n3dfkm461lm3k5hfwxd7biwyd386nk")
+
+    # sunxi: Add arm64 FEL support
+    # https://patchwork.ozlabs.org/patch/1402908
+    (pw "1402908" "1pm6zr7xgafw04rb95237q1mxb8xa6k2lmhzfsj0x0gxal96rw2v")
+
+    # sunxi: dram: h6: Improve DDR3 config detection
+    # https://patchwork.ozlabs.org/patch/1410521
+    (pw "1410521" "1vmalpf3wl3ii592qsh57g11qfrkmx40hlg8mza1330np2xfgb7v")
+
+    # Enable led on boot to notify user of boot status
+    # https://gitlab.com/pine64-org/u-boot/-/commit/1a72bd9f5b0d75361b0852ed515d40a47c1e9bfe
+    (pine64Patch "1a72bd9f5b0d75361b0852ed515d40a47c1e9bfe" "0z6md01qcsf6xwb6hyn4cgxjz2d0dhv05k5hrqp9kwsh9yghchqq")
+
+    # pinephone: Add volume_key environment variable
+    # https://gitlab.com/pine64-org/u-boot/-/commit/2ce71f93c215bf87a9a64a93c3b2c28a86158e5b
+    (pine64Patch "2ce71f93c215bf87a9a64a93c3b2c28a86158e5b" "1x5p0p6zyd4y7yigyikhlwhrywqmd0kpkdnb38583svqzadbr9gw")
```

Comment:
Does this work for the PineTab too?

You will need to use the serial adapter, cancel the boot, and check the u-boot environment variable `volume_key`.

It would be useful to know, as I plan to make use of this in the future with the Pinephone boot setup. The Pinetab would, too, benefit from it.

Reply:
Yes, I confirm that it does work:

```
U-Boot SPL 2021.01 (Jan 11 2021 - 18:11:43 +0000)
DRAM: 2048 MiB
Trying to boot from MMC2
NOTICE:  BL31: v2.4(release):
NOTICE:  BL31: Built : 00:00:00, Jan  1 1980
NOTICE:  BL31: Detected Allwinner A64/H64/R18 SoC (1689)
NOTICE:  BL31: Found U-Boot DTB at 0x208abb8, model: PineTab
NOTICE:  PSCI: System suspend is available via SCPI


U-Boot 2021.01 (Jan 11 2021 - 18:11:43 +0000) Allwinner Technology

CPU:   Allwinner A64 (SUN50I)
Model: PineTab
DRAM:  2 GiB
MMC:   mmc@1c0f000: 0, mmc@1c10000: 2, mmc@1c11000: 1
Loading Environment from FAT... Unable to use mmc 1:3... In:    serial
Out:   serial
Err:   serial
Net:   No ethernet found.
starting USB...
No working controllers found
Hit any key to stop autoboot:  0 
=> printenv volume_key
volume_key=up
```